nexus/tests/integration_tests/password_login.rs

```diff
@@ -0,0 +1,466 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use dropshot::test_util::ClientTestContext;
+use http::{header, method::Method, StatusCode};
+use nexus_passwords::MIN_EXPECTED_PASSWORD_VERIFY_TIME;
+use nexus_test_utils::http_testing::{AuthnMode, NexusRequest, RequestBuilder};
+use nexus_test_utils::resource_helpers::grant_iam;
+use nexus_test_utils::resource_helpers::{create_local_user, create_silo};
+use nexus_test_utils::ControlPlaneTestContext;
+use nexus_test_utils_macros::nexus_test;
+use omicron_common::api::external::Name;
+use omicron_nexus::authz::SiloRole;
+use omicron_nexus::external_api::params;
+use omicron_nexus::external_api::shared;
+use omicron_nexus::external_api::views;
+use std::str::FromStr;
+
+// TODO-coverage verify that deleting a Silo deletes all the users and their
+// password hashes
+
+// TODO-coverage A more rigorous test to verify there are no timing attack
+// vulnerabilities here might be to construct a few kinds of logins (attempt for
+// nonexistent user, attempt for user with no password set, successful attempt,
+// and attempt for user with a password set that doesn't match what's provided).
+// Then run each of these a a bunch of times and verify there's no statistically
+// significant difference between them.
+
+#[nexus_test]
+async fn test_local_users(cptestctx: &ControlPlaneTestContext) {
```

Comment:
I spent a while thinking about this and wrote a bunch about it below and came to the same conclusion: we should just relax this to, say, 900ms.  Unnecessary detail follows.

----

My initial thinking here was:

- It's essential that passwords take a minimum time to be verified on production hardware. (This isn't quite right.  What's essential is that password verification take a minimum time on the hardware that attackers are likely to have access to.  We might consider that "relatively high-end, current-generation hardware".  If attackers happen to be using Gimlets for that, that's coincidental.)
- We should verify this in the test suite (because I can't think of how else we will avoid allowing these parameters to become unsafe as hardware evolves).
- Gimlets are likely to be faster than our dev/test machines.  (This may not be true, and it's irrelevant for the reason mentioned above.)
- So I'll pick parameters that satisfy the timing requirements on Gimlets and hardcode those (both in the implementation and the test suite).  (This makes no sense, for the reason mentioned above.)
- But I can't easily run these tests on Gimlets yet, so I'll pick something that works on my reasonably modern dev machine.  (This was probably dubious before and now makes no sense.)

There's also the fact that if the engineering team uses a diversity of hardware, then choosing safe values here might be painful for people using slower machines.  I'd like to ignore this problem until we confirm it's a problem.  That's because the alternative seems to be introducing a second, less secure set of parameters.  I'm afraid that if we make that configurable, it will be too easy to accidentally use those in production.

So it shouldn't be "we want it to take X time on Gimlets, and we use Y as an approximation and assume it will be Z > Y on all dev and test machines" (which is how I was thinking of it before).  Rather, it's: we want it to take at least X' on all development, test, and production machines, under the assumption that these are likely to be close enough to what attackers are likely to have access to.  There's still the question of what this value should be -- I'll respond to your other comment about that.  The short version is: I think 1s is still a good target, but 900ms is both expedient (because we know it works on both our machines) and still well within the target range described in footnote 2 of RFD 321.

Reply:
Fixed in ec5bfd92 by relaxing the requirement to 900ms.

Reply:
WFM, Thanks!